File: genetic_algorithm.py
import random
import math

POPULATION_SIZE = 100
GENERATIONS = 200
MUTATION_RATE = 0.05
CROSSOVER_RATE = 0.8
ELITE_SIZE = 5

def fitness(chromosome):
    x, y = chromosome
    return 1 / (1 + math.sin(x) * math.cos(y) + 0.1 * (x**2 + y**2))

def create_chromosome():
    return (random.uniform(-10, 10), random.uniform(-10, 10))

def select_parents(population, fitnesses):
    tournament = random.sample(list(zip(population, fitnesses)), 3)
    return max(tournament, key=lambda x: x[1])[0]

def crossover(parent1, parent2):
    if random.random() < CROSSOVER_RATE:
        alpha = random.random()
        child1 = (
            alpha * parent1[0] + (1 - alpha) * parent2[0],
            alpha * parent1[1] + (1 - alpha) * parent2[1]
        )
        child2 = (
            (1 - alpha) * parent1[0] + alpha * parent2[0],
            (1 - alpha) * parent1[1] + alpha * parent2[1]
        )
        return child1, child2
    return parent1, parent2

def mutate(chromosome):
    if random.random() < MUTATION_RATE:
        return (
            chromosome[0] + random.uniform(-0.5, 0.5),
            chromosome[1] + random.uniform(-0.5, 0.5)
        )
    return chromosome

def genetic_algorithm():
    population = [create_chromosome() for _ in range(POPULATION_SIZE)]
    
    for generation in range(GENERATIONS):
        fitnesses = [fitness(chromosome) for chromosome in population]
        
        sorted_pop = sorted(zip(population, fitnesses), key=lambda x: x[1], reverse=True)
        
        print(f"Generation {generation + 1}: Best fitness = {sorted_pop[0][1]:.6f}")
        
        new_population = [chromosome for chromosome, _ in sorted_pop[:ELITE_SIZE]]
        
        while len(new_population) < POPULATION_SIZE:
            parent1 = select_parents(population, fitnesses)
            parent2 = select_parents(population, fitnesses)
            child1, child2 = crossover(parent1, parent2)
            new_population.append(mutate(child1))
            if len(new_population) < POPULATION_SIZE:
                new_population.append(mutate(child2))
        
        population = new_population
    
    best = max(population, key=fitness)
    return best, fitness(best)

if __name__ == "__main__":
    random.seed(42)
    best_solution, best_fitness = genetic_algorithm()
    print(f"\nBest solution: x = {best_solution[0]:.4f}, y = {best_solution[1]:.4f}")
    print(f"Best fitness: {best_fitness:.6f}")

File: genetic_algorithm.js
const POPULATION_SIZE = 100;
const GENERATIONS = 200;
const MUTATION_RATE = 0.05;
const CROSSOVER_RATE = 0.8;
const ELITE_SIZE = 5;

function random(min, max) {
    return Math.random() * (max - min) + min;
}

function fitness(chromosome) {
    const [x, y] = chromosome;
    return 1 / (1 + Math.sin(x) * Math.cos(y) + 0.1 * (x * x + y * y));
}

function createChromosome() {
    return [random(-10, 10), random(-10, 10)];
}

function selectParents(population, fitnesses) {
    const tournament = [];
    for (let i = 0; i < 3; i++) {
        const idx = Math.floor(Math.random() * population.length);
        tournament.push({ chromosome: population[idx], fit: fitnesses[idx] });
    }
    return tournament.reduce((max, p) => p.fit > max.fit ? p : max).chromosome;
}

function crossover(parent1, parent2) {
    if (Math.random() < CROSSOVER_RATE) {
        const alpha = Math.random();
        const child1 = [
            alpha * parent1[0] + (1 - alpha) * parent2[0],
            alpha * parent1[1] + (1 - alpha) * parent2[1]
        ];
        const child2 = [
            (1 - alpha) * parent1[0] + alpha * parent2[0],
            (1 - alpha) * parent1[1] + alpha * parent2[1]
        ];
        return [child1, child2];
    }
    return [parent1, parent2];
}

function mutate(chromosome) {
    if (Math.random() < MUTATION_RATE) {
        return [
            chromosome[0] + random(-0.5, 0.5),
            chromosome[1] + random(-0.5, 0.5)
        ];
    }
    return chromosome;
}

function geneticAlgorithm() {
    let population = Array.from({ length: POPULATION_SIZE }, createChromosome);
    
    for (let generation = 0; generation < GENERATIONS; generation++) {
        const fitnesses = population.map(fitness);
        
        const sorted = population
            .map((c, i) => [c, fitnesses[i]])
            .sort((a, b) => b[1] - a[1]);
        
        console.log(`Generation ${generation + 1}: Best fitness = ${sorted[0][1].toFixed(6)}`);
        
        let newPopulation = sorted.slice(0, ELITE_SIZE).map(x => x[0]);
        
        while (newPopulation.length < POPULATION_SIZE) {
            const parent1 = selectParents(population, fitnesses);
            const parent2 = selectParents(population, fitnesses);
            const [child1, child2] = crossover(parent1, parent2);
            newPopulation.push(mutate(child1));
            if (newPopulation.length < POPULATION_SIZE) {
                newPopulation.push(mutate(child2));
            }
        }
        
        population = newPopulation;
    }
    
    const best = population.reduce((best, c) => fitness(c) > fitness(best) ? c : best);
    return [best, fitness(best)];
}

const [bestSolution, bestFitness] = geneticAlgorithm();
console.log(`\nBest solution: x = ${bestSolution[0].toFixed(4)}, y = ${bestSolution[1].toFixed(4)}`);
console.log(`Best fitness: ${bestFitness.toFixed(6)}`);